File: Examples/intel_examples/examples_core_c/vmlc/source/vzexp.c
/*******************************************************************************
* Copyright 2001-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    vzExp  Example Program Text
!******************************************************************************/

#include <stdio.h>
#include "mkl_vml.h"

#include "_rms.h"

#define INCA 3
#define INCB 5

int main()
{
  MKL_Complex16 zA[10],zB[10];
  MKL_Complex16 zBha0[10],zBha1[10],zBha2[10];
  MKL_Complex16           zBla1[10],zBla2[10];
  MKL_Complex16           zBep1[10],zBep2[10];
  float CurRMS,MaxRMS=0.0;

  MKL_Complex16 zA_I[10*INCA],zB_I[10*INCB];
  MKL_Complex16 zBha0_I[10*INCB],zBha1_I[10*INCB],zBha2_I[10*INCB];
  MKL_Complex16                  zBla1_I[10*INCB],zBla2_I[10*INCB];
  MKL_Complex16                  zBep1_I[10*INCB],zBep2_I[10*INCB];
  float CurRMS_I,MaxRMS_I=0.0;
  MKL_INT inca=INCA,incb=INCB;

  MKL_INT i=0,vec_len=10;

  zA[0].real=-17.0000;zA[0].imag=18.0000;
  zA[1].real=-13.1111;zA[1].imag=14.1111;
  zA[2].real=-9.2222;zA[2].imag=10.2222;
  zA[3].real=-5.3333;zA[3].imag=6.3333;
  zA[4].real=-1.4444;zA[4].imag=2.4444;
  zA[5].real=2.4444;zA[5].imag=-1.4444;
  zA[6].real=6.3333;zA[6].imag=-5.3333;
  zA[7].real=10.2222;zA[7].imag=-9.2222;
  zA[8].real=14.1111;zA[8].imag=-13.1111;
  zA[9].real=18.0000;zA[9].imag=-17.0000;
  zB[0].real=2.7336700468037275e-008;zB[0].imag=-3.1090404292366861e-008;
  zB[1].real=5.2718434590867303e-008;zB[1].imag=2.0219669530527123e-006;
  zB[2].real=-6.9031805334250157e-005;zB[2].imag=-7.0712150283650660e-005;
  zB[3].real=4.8220493390719000e-003;zB[3].imag=2.4185802936554069e-004;
  zB[4].real=-1.8084266353749434e-001;zB[4].imag=1.5145585212254198e-001;
  zB[5].real=1.4526697215961737e+000;zB[5].imag=-1.1431704865446497e+001;
  zB[6].real=3.2754677379631244e+002;zB[6].imag=4.5792469439319871e+002;
  zB[7].real=-2.6944627012817422e+004;zB[7].imag=-5.5343007569010069e+003;
  zB[8].real=1.1494097020286692e+006;zB[8].imag=-6.9640053097577486e+005;
  zB[9].real=-1.8067216284192696e+007;zB[9].imag=6.3125329645518661e+007;

  for(i=0;i<10;i++) {
    zA_I[i*inca]=zA[i];
    zB_I[i*incb]=zB[i];
  }

  vzExp(vec_len,zA,zBha0);
  vzExpI(vec_len,zA_I,inca,zBha0_I,incb);

  vmzExp(vec_len,zA,zBep1,VML_EP);
  vmzExpI(vec_len,zA_I,inca,zBep1_I,incb,VML_EP);

  vmlSetMode(VML_EP);
  vzExp(vec_len,zA,zBep2);
  vzExpI(vec_len,zA_I,inca,zBep2_I,incb);

  vmzExp(vec_len,zA,zBla1,VML_LA);
  vmzExpI(vec_len,zA_I,inca,zBla1_I,incb,VML_LA);

  vmlSetMode(VML_LA);
  vzExp(vec_len,zA,zBla2);
  vzExpI(vec_len,zA_I,inca,zBla2_I,incb);

  vmzExp(vec_len,zA,zBha1,VML_HA);
  vmzExpI(vec_len,zA_I,inca,zBha1_I,incb,VML_HA);

  vmlSetMode(VML_HA);
  vzExp(vec_len,zA,zBha2);
  vzExpI(vec_len,zA_I,inca,zBha2_I,incb);

  for(i=0;i<10;i++) {
    if(zBha0[i].real!=zBha1[i].real || zBha0[i].imag!=zBha1[i].imag || zBha1[i].real!=zBha2[i].real ||
        zBha1[i].imag!=zBha2[i].imag) {
      printf("Error! Difference between vzExp and vmzExp in VML_HA mode detected.\n");
      return 1;
    }

    if(zBha0_I[i*incb].real!=zBha1_I[i*incb].real || zBha0_I[i*incb].imag!=zBha1_I[i*incb].imag ||
        zBha1_I[i*incb].real!=zBha2_I[i*incb].real || zBha1_I[i*incb].imag!=zBha2_I[i*incb].imag) {
      printf("Error! Difference between vzExpI and vmzExpI in VML_HA mode detected.\n");
      return 1;
    }

    if(zBla1[i].real!=zBla2[i].real || zBla1[i].imag!=zBla2[i].imag) {
      printf("Error! Difference between vzExp and vmzExp in VML_LA mode detected.\n");
      return 1;
    }

    if(zBla1_I[i*incb].real!=zBla2_I[i*incb].real || zBla1_I[i*incb].imag!=zBla2_I[i*incb].imag) {
      printf("Error! Difference between vzExpI and vmzExpI in VML_LA mode detected.\n");
      return 1;
    }

    if(zBep1[i].real!=zBep2[i].real || zBep1[i].imag!=zBep2[i].imag) {
      printf("Error! Difference between vzExp and vmzExp in VML_EP mode detected.\n");
      return 1;
    }

    if(zBep1_I[i*incb].real!=zBep2_I[i*incb].real || zBep1_I[i*incb].imag!=zBep2_I[i*incb].imag) {
      printf("Error! Difference between vzExpI and vmzExpI in VML_EP mode detected.\n");
      return 1;
    }
  }

  printf("vzExp test/example program\n\n");
  printf("           Argument                           vzExp\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f %+.10f*i\n",zA[i].real,zA[i].imag,zBha0[i].real,zBha0[i].imag);
    CurRMS=zrelerr(zB[i],zBha0[i]);
    if(CurRMS>MaxRMS) MaxRMS=CurRMS;
  }
  printf("\n");
  printf("vzExpI test/example program\n\n");
  printf("           Argument                           vzExp\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f %+.10f*i\n",zA_I[i*inca].real,zA_I[i*inca].imag,zBha0_I[i*incb].real,
        zBha0_I[i*incb].imag);
    CurRMS_I=zrelerr(zB_I[i*incb],zBha0_I[i*incb]);
    if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I;
  }
  printf("\n");
  if(MaxRMS>=1e-5) {
    printf("Error! Relative accuracy is %.16f\n",MaxRMS);
    return 1;
  }
  else {
    printf("Relative accuracy is %.16f\n",MaxRMS);
  }

  printf("\n");

  if(MaxRMS_I>=1e-5) {
    printf("Error! Relative strided accuracy is %.16f\n",MaxRMS_I);
    return 1;
  }
  else {
    printf("Relative strided accuracy is %.16f\n",MaxRMS_I);
  }
  return 0;
}

File: Examples/intel_examples/examples_core_f/vmlf/source/vzexp.f
!===============================================================================
! Copyright 2001-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    vzExp  Example Program Text
!*******************************************************************************

      include "_rms.fi"

      program MKL_VML_TEST

      include "mkl_vml.f90"
      real(kind=4) :: srelerr
      real(kind=8) :: drelerr
      real(kind=4) :: crelerr
      real(kind=8) :: zrelerr

      complex(kind=8) zA(10)
      complex(kind=8) zB(10)
      complex(kind=8) zBha0(10)
      complex(kind=8) zBha1(10)
      complex(kind=8) zBha2(10)
      complex(kind=8) zBla1(10)
      complex(kind=8) zBla2(10)
      complex(kind=8) zBep1(10)
      complex(kind=8) zBep2(10)

      real(kind=8) CurRMS,MaxRMS

      integer, parameter :: inca=3
      integer, parameter :: incb=5
      complex(kind=8) zA_I(10*inca)
      complex(kind=8) zB_I(10*incb)
      complex(kind=8) zBha0_I(10*incb)
      complex(kind=8) zBha1_I(10*incb)
      complex(kind=8) zBha2_I(10*incb)
      complex(kind=8) zBla1_I(10*incb)
      complex(kind=8) zBla2_I(10*incb)
      complex(kind=8) zBep1_I(10*incb)
      complex(kind=8) zBep2_I(10*incb)
      real(kind=8) CurRMS_I,MaxRMS_I

      integer(kind=8) mode
      integer tmode
      integer i, vec_len

      vec_len=10
      MaxRMS=0.0
      MaxRMS_I=0.0

      zA( 1)=(-17.0000d+000,18.0000d+000)
      zA( 2)=(-13.1111d+000,14.1111d+000)
      zA( 3)=(-9.2222d+000,10.2222d+000)
      zA( 4)=(-5.3333d+000,6.3333d+000)
      zA( 5)=(-1.4444d+000,2.4444d+000)
      zA( 6)=(2.4444d+000,-1.4444d+000)
      zA( 7)=(6.3333d+000,-5.3333d+000)
      zA( 8)=(10.2222d+000,-9.2222d+000)
      zA( 9)=(14.1111d+000,-13.1111d+000)
      zA(10)=(18.0000d+000,-17.0000d+000)
      zB( 1)=(2.7336700468037275d-008,-3.1090404292366861d-008)
      zB( 2)=(5.2718434590867303d-008,2.0219669530527123d-006)
      zB( 3)=(-6.9031805334250157d-005,-7.0712150283650660d-005)
      zB( 4)=(4.8220493390719000d-003,2.4185802936554069d-004)
      zB( 5)=(-1.8084266353749434d-001,1.5145585212254198d-001)
      zB( 6)=(1.4526697215961737d+000,-1.1431704865446497d+001)
      zB( 7)=(3.2754677379631244d+002,4.5792469439319871d+002)
      zB( 8)=(-2.6944627012817422d+004,-5.5343007569010069d+003)
      zB( 9)=(1.1494097020286692d+006,-6.9640053097577486d+005)
      zB(10)=(-1.8067216284192696d+007,6.3125329645518661d+007)

      do i=1,10
          zA_I((i-1)*inca+1)=zA(i)
          zB_I((i-1)*incb+1)=zB(i)
      end do

      call VZEXP(vec_len,zA,zBha0)
      call VZEXPI(vec_len,zA_I,inca,zBha0_I,incb)

      mode=VML_EP
      call VMZEXP(vec_len,zA,zBep1,mode)
      call VMZEXPI(vec_len,zA_I,inca,zBep1_I,incb,mode)

      tmode=VML_EP
      tmode=VMLSETMODE(tmode)
      call VZEXP(vec_len,zA,zBep2)
      call VZEXPI(vec_len,zA_I,inca,zBep2_I,incb)

      mode=VML_LA
      call VMZEXP(vec_len,zA,zBla1,mode)
      call VMZEXPI(vec_len,zA_I,inca,zBla1_I,incb,mode)

      tmode=VML_LA
      tmode=VMLSETMODE(tmode)
      call VZEXP(vec_len,zA,zBla2)
      call VZEXPI(vec_len,zA_I,inca,zBla2_I,incb)

      mode=VML_HA
      call VMZEXP(vec_len,zA,zBha1,mode)
      call VMZEXPI(vec_len,zA_I,inca,zBha1_I,incb,mode)

      tmode=VML_HA
      tmode=VMLSETMODE(tmode)
      call VZEXP(vec_len,zA,zBha2)
      call VZEXPI(vec_len,zA_I,inca,zBha2_I,incb)

      do i=1,10
          if(zBha0(i) .ne. zBha1(i)) then
              print *,"Error! Difference between VZEXP and"
              print *," VMZEXP in VML_HA mode detected"
              stop 1
          endif
          if(zBha0_I((i-1)*incb+1) .ne. zBha1_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZEXPI and"
              print *," VMZEXPI in VML_HA mode detected"
              stop 1
          endif
          if(zBha1(i) .ne. zBha2(i)) then
              print *,"Error! Difference between VZEXP and"
              print *," VMZEXP in VML_HA mode detected"
              stop 1
          endif
          if(zBha1_I((i-1)*incb+1) .ne. zBha2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZEXPI and"
              print *," VMZEXPI in VML_HA mode detected"
              stop 1
          endif
          if(zBla1(i) .ne. zBla2(i)) then
              print *,"Error! Difference between VZEXP and"
              print *," VMZEXP in VML_LA mode detected"
              stop 1
          endif
          if(zBla1_I((i-1)*incb+1) .ne. zBla2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZEXPI and"
              print *," VMZEXPI in VML_LA mode detected"
              stop 1
          endif
          if(zBep1(i) .ne. zBep2(i)) then
              print *,"Error! Difference between VZEXP and"
              print *," VMZEXP in VML_EP mode detected"
              stop 1
          endif
          if(zBep1_I((i-1)*incb+1) .ne. zBep2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZEXPI and"
              print *," VMZEXPI in VML_EP mode detected"
              stop 1
          endif
      end do

      print *,"vzExp test/example program"
      print *,""
      print *,"           Argument                     vzExp"
      print *,"======================================================", &
     &        "========================"
      do i=1,vec_len
            print 10,zA(i),"    ",zBha0(i)
            CurRMS=zrelerr(zB(i),zBha0(i))
            if(CurRMS>MaxRMS) MaxRMS=CurRMS
      end do

      print *,""
      print *,"vzExpI test/example program"
      print *,"           Argument                     vzExp"
      print *,"======================================================", &
     &        "========================"
      do i=1,10
            print 10,zA_I((i-1)*inca+1),"    ",zBha0_I((i-1)*incb+1)
            CurRMS_I=zrelerr(zB_I((i-1)*incb+1),zBha0_I((i-1)*incb+1))
            if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I
      end do

      print *,""
      if(MaxRMS>=1e-14) then
            print 11,"Error! Relative accuracy is ",MaxRMS
            stop 1
      else
            print 11,"Relative accuracy is ",MaxRMS
      endif

      if(MaxRMS_I>=1e-14) then
            print 11,"Error! Relative strided accuracy is ",MaxRMS_I
            stop 1
      else
            print 11,"Relative strided accuracy is ",MaxRMS_I
      endif

10    format(E15.7,E15.7,A5,E15.7,E15.7)
11    format(A,F25.16)

      end
